### ISSUE TITLE
Fix for non-static bad pixels

### ISSUE THREAD

**Issue by seescho** [ilia3101/MLV-App#61]
Is it very difficult to add a fix for non-static bad pixels into mlvapp?

mlv_dump has such an option (--fixcp2). But only cold-pixels are fixed there using the "find_and_fix_cold_pixels"-function, Alex and I coded some years ago into raw2dng. In mlvapp, hot-pixels are repaired as well, but only the static-ones.

I love the bad-pixel fix in mlvapp. But for my low-light and high-ISO astro-stuff, a possibility to remove non-static bad-pixels would be great. Decoding-speed would go down with such an option enabled, because every single video-frame would have be analyzed. But this price is not to high, I guess.

**Comment by bouncyball-git** [comment]
I know, I dumped this feature in mlv_dump on steroids and used the one now implemented to mlvapp (mlvfs style). 

On the other hand I've implemented every frame stripe correction b/c of someone asked on the forum.

Of course it is not very difficult to do what you need. We're gonna need to change gui:

off | on | aggressive | force | force aggresive

or make force aggressive by default?

**Comment by bouncyball-git** [comment]
or maybe like this

off | on | firce
default | aggressive

@masc4ii: what do you think?

**Comment by seescho** [comment]
"off | on | force
default | aggressive"
would be fine with defaults to "on" and "default"
aggressive as default is not the best, because that method has artefacts in some cases.

**Comment by bouncyball-git** [comment]
@masc4ii: what do you think? Should we make happy our German Jury? ;)

**Comment by bouncyball-git** [comment]
off | **on** | force
**normal** | aggresive

Is the goal.

**Comment by masc4ii** [comment]
Let's do it.
Fix Bad Pixels
| off | on | force |
Search Method | normal | aggressive |
Interpolation Method | 1 | 2 |
We only should think about making the names a bit shorter...

**Comment by seescho** [comment]
One question about the interpolation mode.

We have Method 1 and Method 2. I think, most of the users don´t know, what´s the difference between these two methods.

Yesterday, I had some time (a nice snowstorm). I played with the bad-pixel-stuff. But all what I did, I did not see any difference in the results of the two methods.

Is it necessary to have these two interpolation-methods?

**Comment by bouncyball-git** [comment]
The difference is more apparent with focus pixels. If you have some 100D (has lots of them) clips try and you'll see yourself.

Also there is a real chance to get bad artifacts in conjunction with chroma smooth. With method 2 CS works better.

**Comment by bouncyball-git** [comment]
Me and Danne been experimenting with this a while ago and decided to have both.

**Comment by bouncyball-git** [comment]
called 'llrawproc->bps_method'. Bad Pixel Search Method

**Comment by bouncyball-git** [comment]
Commited change for testing. 

At the moment works like this:
1. If bad pixels on - searches bad pixels for only 1st frame with default search method
2. If bad pixels aggressive - searches bad pixels for every frame with default search method

Search methid will be added when @masc4ii implements this to gui :)

**Comment by masc4ii** [comment]
I think we have a small mistake here. Before we had:
Fix Bad Pixels
| off | **on** | aggressive |
Interpolation Method | 1 | **2** |

So now I'll commit:
Fix Bad Pixels
| off | **on** | aggressive |
Search Method | **normal** | force |
Interpolation Method | 1 | **2** |

Else we should rename many many functions, elements, comments, etc. ...

**Comment by seescho** [comment]
`The difference is more apparent with focus pixels...`
I see.

`Committed change for testing...`
Tested. Looks fine. :)

OT: I guess, it´s time for some new screenshoots for my linux-compile-howto. Must look, what I can do here the next few days ;)

**Comment by bouncyball-git** [comment]
Last commit for this topic pushed. Please test and if everything is ok let's close this issue :)

**Comment by bouncyball-git** [comment]
@masc4ii: I renamed all needed stuff plus made signal->slot connection for new toolbuttons, everything is working so far. My 1st signal slot journey :)

**Comment by masc4ii** [comment]
Haha... yeay... and I forgot to connections! Well done! ;-)

**Comment by bouncyball-git** [comment]
Thank you :)

**Comment by bouncyball-git** [comment]
@masc4ii 

OT: Can active toolbuttons in raw correction section be made a bit lighter gray and the foregroung text color black/dark gray? Like in cocoa version? It is really more appealing.

**Comment by masc4ii** [comment]
I can change the backgroundcolor, but then the text isn't there when the button is inactive (overlay grey). The text would be there, if it is black, yes. Unfortunately Qt does not offer a "highlighted text", so we would have to change Qt libraries here (I have no idea how difficult it is in this case)... 
In past I did not change here after having some trouble. And I liked more the Qt style than the cocoa style... but it seems this is a question of taste ;-)

**Comment by seescho** [comment]
Thank You all for implementing this fix. I love it.

**Comment by bouncyball-git** [comment]
U R Welcome :)

Is it working as expected?

**Comment by seescho** [comment]
I tested all modes and it worked as expected, so far I could see.

Here the exporting-times for all modes:
```
off 26s
on-normal 27s
on-agressive 27s

force-normal 35s
force-aggresive 62s
```
video-data:
```
1762 frames
704px x 396px
14bit umcompressed

exported to tiff
```
The static bad-pixels are not so easy to test, because the non-static-ones are much much more present in my records. I loaded some frames into GIMP (subtraction-mode) to see, what happens with the static bad-pixels. Looks like all is fine. But this needs more testing to be sure (maybe later this year :) ).

Force mode works great with both, normal and aggresive mode.

Just a few bad pixels are not removed in aggressive mode, bur I´m not sure, they really are bad-pixels (Will analyze this maybe later this year, too :) )

All in all: Great enhancement!
